### PR TITLE
UAR-712: Null pointer exception during an update submission when there is no MO data for an OE

### PR DESCRIPTION
**Pull request by bbell-companieshouse**
### Work checklist

- [ ] Tests added where applicable

### Merge instructions

We are committed to keeping commit history clean, consistent and linear. To achieve this, this commit should be structured as follows:

```
<type>[optional scope]: <description>
```

and contain the following structural elements:

- fix: a commit that patches a bug in your codebase (this correlates with PATCH in semantic versioning),
- feat: a commit that introduces a new feature to the codebase (this correlates with MINOR in semantic versioning),
- BREAKING CHANGE: a commit that has a footer `BREAKING CHANGE:` introduces a breaking API change (correlating with MAJOR in semantic versioning). A BREAKING CHANGE can be part of commits of any type,
- types other than `fix:` and `feat:` are allowed, for example `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`, and others,
- footers other than `BREAKING CHANGE: <description>` may be provided.

### CODE REVIEW

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/utils/PublicPrivateDataCombiner.java [hunk 1]
```diff
@@ -63,42 +63,44 @@ public Map<String, Pair<PscApi, PrivateBoDataApi>> buildMergedBeneficialOwnerDat
...
-
-      String hashedId = null;
-      try {
-        hashedId = hashHelper.encode(privateBO.getPscId());
-      } catch (NoSuchAlgorithmException e) {
-        throw new ServiceException("Error hashing PSC ID", e);
+    if (privateBOs != null && publicPSCs != null) {
+      for (PrivateBoDataApi privateBO : privateBOs.getBoPrivateData()) {
+
+        String hashedId = null;
```
Comment (jkayyala) [inline src/main/java/uk/gov/companieshouse/overseasentitiesapi/utils/PublicPrivateDataCombiner.java:69]: Can't this be extracted into a method 
```java
private String retrieveHashedId(String officerId) throws ServiceException {
    String hashedId = null;
    try {
      hashedId = hashHelper.encode(officerId);
    } catch (NoSuchAlgorithmException e) {
      throw new ServiceException("Error hashing Managing Officer ID", e);
    }
    return hashedId;
  }
```

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/utils/PopulateUpdateSubmission.java [hunk 2]
```diff
@@ -29,6 +29,7 @@ public class PopulateUpdateSubmission {
      */
     public void populate(OverseasEntitySubmissionDto overseasEntitySubmissionDto,
             UpdateSubmission updateSubmission) {
+        updateSubmission.setEntityNumber(overseasEntitySubmissionDto.getEntityNumber());
```
Comment (n-r-b) [inline src/main/java/uk/gov/companieshouse/overseasentitiesapi/utils/PopulateUpdateSubmission.java:32]: Good idea moving this, can you update the tests just to confirm this is being set?

File: src/test/java/uk/gov/companieshouse/overseasentitiesapi/utils/PublicPrivateDataCombinerTest.java [hunk 3]
```diff
@@ -312,4 +313,104 @@ void testLogCollatedDataWrongHash() throws ServiceException, JsonProcessingExcep
 
     assertEquals(expectedOutput, actualOutput);
   }
+
+  @Test
+  void testBuildMergedBeneficialOwnerDataMapNoPrivatePublicBoData() throws JsonProcessingException, ServiceException {
```
Comment (n-r-b) [inline src/test/java/uk/gov/companieshouse/overseasentitiesapi/utils/PublicPrivateDataCombinerTest.java:318]:
```suggestion
  void testBuildMergedBeneficialOwnerDataMapNoPrivatePublicBoData() throws ServiceException {
```

File: src/test/java/uk/gov/companieshouse/overseasentitiesapi/utils/PublicPrivateDataCombinerTest.java [hunk 4]
```diff
@@ -312,4 +313,104 @@ void testLogCollatedDataWrongHash() throws ServiceException, JsonProcessingExcep
...
+    // Prepare test data
+    ManagingOfficerDataApi managingOfficerDataApi = new ManagingOfficerDataApi();
+    managingOfficerDataApi.setManagingOfficerId("12345");
+
+    OfficersApi officersApi = objectMapper.readValue(officersApiString, OfficersApi.class);
+    CompanyOfficerApi companyOfficerApi = officersApi.getItems().get(0);
+
+    // Configure mock behavior
+    when(privateDataRetrievalService.getManagingOfficerData()).thenReturn(null);
+    when(publicDataRetrievalService.getOfficers()).thenReturn(null);
```
Comment (n-r-b) [inline src/test/java/uk/gov/companieshouse/overseasentitiesapi/utils/PublicPrivateDataCombinerTest.java:377]: Do these mocks needs to be updated with MO data? Potentially using the objects created above

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/utils/PopulateUpdateSubmission.java (hunk 2 above)
Comment (bbell-companieshouse) [reply src/main/java/uk/gov/companieshouse/overseasentitiesapi/utils/PopulateUpdateSubmission.java:32]: I believe its already tested in testFilingGenerationWhenSuccessfulWithoutTrustsAndWithIdentityChecksForUpdate in FilingServiceTest

File: src/test/java/uk/gov/companieshouse/overseasentitiesapi/utils/PublicPrivateDataCombinerTest.java (hunk 4 above)
Comment (bbell-companieshouse) [reply src/test/java/uk/gov/companieshouse/overseasentitiesapi/utils/PublicPrivateDataCombinerTest.java:366]: Forgot to remove those mocks- good spot :D

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/utils/PublicPrivateDataCombiner.java [hunk 5]
```diff
@@ -3,258 +3,280 @@
...
+                pairToPutInMap = Pair.of(null, privateMO);
+            }
+
+            combinedMOs.put(hashedId, pairToPutInMap);
+        }
+    }
+
+    private void putPublicMoData(OfficersApi publicOfficers) {
+
+        for (CompanyOfficerApi publicOfficer : publicOfficers.getItems()) {
```
Comment (n-r-b) [inline src/main/java/uk/gov/companieshouse/overseasentitiesapi/utils/PublicPrivateDataCombiner.java:252]: Can items be null? The exception was on this line but I'm not sure if it was due to `publicOfficers` being null or `getItems` returning null

Comment (n-r-b) [reply src/main/java/uk/gov/companieshouse/overseasentitiesapi/utils/PublicPrivateDataCombiner.java:252]: Looking at PublicDataRetrievalService.java, line 86, if the officer data cannot be found for a company number we return an `officersApi` object but only set the `totalResults` so the `items` data may be null.

File: src/test/java/uk/gov/companieshouse/overseasentitiesapi/utils/PublicPrivateDataCombinerTest.java [hunk 6]
```diff
@@ -312,4 +313,93 @@ void testLogCollatedDataWrongHash() throws ServiceException, JsonProcessingExcep
...
+    Map<String, Pair<CompanyOfficerApi, ManagingOfficerDataApi>> results = publicPrivateDataCombiner.buildMergedManagingOfficerDataMap();
+
+    assertTrue(results.isEmpty());
+  }
+
+  @Test
+  void testBuildMergedManagingOfficerDataMapNoPrivateMoData() throws JsonProcessingException, ServiceException {
+    // Prepare test data
+    OfficersApi officersApi = objectMapper.readValue(officersApiString, OfficersApi.class);
+    CompanyOfficerApi companyOfficerApi = officersApi.getItems().get(0);
```
Comment (n-r-b) [inline src/test/java/uk/gov/companieshouse/overseasentitiesapi/utils/PublicPrivateDataCombinerTest.java:378]: This line is not used so can be removed

File: src/test/java/uk/gov/companieshouse/overseasentitiesapi/utils/PublicPrivateDataCombinerTest.java [hunk 7]
```diff
@@ -312,4 +313,93 @@ void testLogCollatedDataWrongHash() throws ServiceException, JsonProcessingExcep
...
+
+    // Configure mock behavior
+    when(privateDataRetrievalService.getManagingOfficerData()).thenReturn(null);
+    when(publicDataRetrievalService.getOfficers()).thenReturn(officersApi);
```
Comment (n-r-b) [inline src/test/java/uk/gov/companieshouse/overseasentitiesapi/utils/PublicPrivateDataCombinerTest.java:382]: Could be simplified
```suggestion
    when(publicDataRetrievalService.getOfficers()).thenReturn(new OfficersApi());
```

File: src/test/java/uk/gov/companieshouse/overseasentitiesapi/utils/PublicPrivateDataCombinerTest.java [hunk 8]
```diff
@@ -312,4 +313,93 @@ void testLogCollatedDataWrongHash() throws ServiceException, JsonProcessingExcep
...
+
+    // Execute the method to test
+    Map<String, Pair<CompanyOfficerApi, ManagingOfficerDataApi>> results = publicPrivateDataCombiner.buildMergedManagingOfficerDataMap();
+
+    assertTrue(results.isEmpty());
+  }
+
+  @Test
+  void testBuildMergedManagingOfficerDataMapNoPublicMoData() throws JsonProcessingException, ServiceException {
```
Comment (n-r-b) [inline src/test/java/uk/gov/companieshouse/overseasentitiesapi/utils/PublicPrivateDataCombinerTest.java:391]: We can clean up the `throws JsonProcessingException` from a few of the new test cases as it's not used

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/utils/PublicPrivateDataCombiner.java [hunk 9]
```diff
@@ -3,258 +3,280 @@
...
+            } else {
+                pairToPutInMap = Pair.of(null, privateBO);
+            }
+            combinedBOs.put(hashedId, pairToPutInMap);
+        }
+    }
 
-    result.append("}");
+    private void putPublicBoData(PscsApi publicPSCs) {
+        for (PscApi publicPSC : publicPSCs.getItems()) {
```
Comment (AndreKramer7654) [inline src/main/java/uk/gov/companieshouse/overseasentitiesapi/utils/PublicPrivateDataCombiner.java:214]: Can getItems() return null?

File: src/test/java/uk/gov/companieshouse/overseasentitiesapi/utils/PublicPrivateDataCombinerTest.java (hunk 7 above)
Comment (bbell-companieshouse) [reply src/test/java/uk/gov/companieshouse/overseasentitiesapi/utils/PublicPrivateDataCombinerTest.java:398]: In this example I wanted to use the same public data used in other tests so i've kept it this way for consistency

**Review by AndreKramer7654 — APPROVED** [review]

**Pull request merged**